dumper/app/src/main/java/com/google/edwmigration/dumper/application/dumper/TasksRunner.java

```diff
@@ -0,0 +1,178 @@
+import com.google.edwmigration.dumper.application.dumper.io.OutputHandleFactory;
+import com.google.edwmigration.dumper.application.dumper.task.Task;
+import com.google.edwmigration.dumper.application.dumper.task.TaskGroup;
+import com.google.edwmigration.dumper.application.dumper.task.TaskRunContext;
+import com.google.edwmigration.dumper.application.dumper.task.TaskSetState;
+import com.google.edwmigration.dumper.application.dumper.task.TaskSetState.Impl;
+import com.google.edwmigration.dumper.application.dumper.task.TaskState;
+import java.nio.charset.StandardCharsets;
+import java.sql.SQLException;
+import java.util.Arrays;
+import java.util.List;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicInteger;
+import javax.annotation.CheckForNull;
+import javax.annotation.Nonnull;
+import org.apache.commons.lang3.time.DurationFormatUtils;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/** @author ishmum */
+public class TasksRunner {
+
+  private static final Logger LOG = LoggerFactory.getLogger(TasksRunner.class);
+  private static final Logger PROGRESS_LOG = LoggerFactory.getLogger("progress-logger");
+
+  private AtomicInteger atomicNumberOfCompletedTasks;
+  private final int totalNumberOfTasks;
+  private final Stopwatch stopwatch;
+
+  private final TaskRunContext context;
+  private final TaskSetState.Impl state;
+  private final List<Task<?>> tasks;
+
+  public TasksRunner(
+      OutputHandleFactory sinkFactory,
+      Handle handle,
+      int threadPoolSize,
+      @Nonnull TaskSetState.Impl state,
+      List<Task<?>> tasks) {
+    context = createContext(sinkFactory, handle, threadPoolSize, state);
+    this.state = state;
+    this.tasks = tasks;
+    totalNumberOfTasks = countTasks(tasks);
+    stopwatch = Stopwatch.createStarted();
+    atomicNumberOfCompletedTasks = new AtomicInteger();
+  }
+
+  private TaskRunContext createContext(
+      OutputHandleFactory sinkFactory, Handle handle, int threadPoolSize, Impl state) {
+    return new TaskRunContext(sinkFactory, handle, threadPoolSize) {
+      @Override
+      public TaskState getTaskState(Task<?> task) {
+        return state.getTaskState(task);
+      }
+
+      @Override
+      public <T> T runChildTask(Task<T> task) throws MetadataDumperUsageException {
+        return handleTask(task);
+      }
+    };
+  }
+
+  public void run() throws MetadataDumperUsageException {
+    for (Task<?> task : tasks) {
+      handleTask(task);
+    }
+  }
+
+  @CheckForNull
+  private <T> T handleTask(Task<T> task) throws MetadataDumperUsageException {
+    T t = runTask(task);
+    if (!(task instanceof TaskGroup)) {
+      atomicNumberOfCompletedTasks.getAndIncrement();
+    }
+    logProgress();
+    return t;
+  }
+
+  private void logProgress() {
+    int numberOfCompletedTasks = atomicNumberOfCompletedTasks.get();
+
+    long averageTimePerTaskInMillisecond =
+        stopwatch.elapsed(TimeUnit.MILLISECONDS) / numberOfCompletedTasks;
```

Comment:
I know this is unlikely with the current implementation, but I would still add protection against division by zero, e.g.

```
numberOfCompletedTasks == 0
  ? 0
  : stopwatch.elapsed(TimeUnit.MILLISECONDS) / numberOfCompletedTasks
```